cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java

```diff
@@ -70,6 +73,208 @@ public LineageOperationsProcessor(Set<Connection> stageConnections,
+          case READ:
+            PipelineReadOperation read = (PipelineReadOperation) pipelineOperation;
+            processInputOutputs(read.getName(), Collections.emptyList(), read.getOutputFields(), validOperationInputs,
+                                invalidInputs, unusedOperationOutputs, redundantOutputs);
+            break;
+          case TRANSFORM:
+            PipelineTransformOperation transform = (PipelineTransformOperation) pipelineOperation;
+            processInputOutputs(transform.getName(), transform.getInputFields(), transform.getOutputFields(),
+                                validOperationInputs, invalidInputs, unusedOperationOutputs, redundantOutputs);
+            break;
+          case WRITE:
+            PipelineWriteOperation write = (PipelineWriteOperation) pipelineOperation;
+            processInputOutputs(write.getName(), write.getInputFields(), Collections.emptyList(), validOperationInputs,
+                                invalidInputs, unusedOperationOutputs, redundantOutputs);
+            break;
+        }
+      }
+
+      // Remove outputs from unusedOperationOutputs map which are not used by any operation as an input,
+      // however they are part of output schema. We cannot simply remove all keys here by doing
+      // "unusedOperationOutputs.removeAll(stageOutputFields.get(stageName))". Reason for this is
+      // consider the case where OP1 outputs field b, and OP2 outputs the same field b. Operation OP1 is
+      // redundant, since b will be overwritten by OP2. However since field b is  not consumed by any other
+      // operation as an input, redundantOutputs will not have its entry. We still want to fail the validation
+      // for operation OP1.
+      Iterator<Map.Entry<String, List<String>>> iterator = unusedOperationOutputs.entrySet().iterator();
+      while (iterator.hasNext()) {
+        Map.Entry<String, List<String>> next = iterator.next();
+        String field = next.getKey();
+        List<String> origins = next.getValue();
+
+        if (stageOutputFields.get(stageName).contains(field)) {
+          iterator.remove();
+          if (origins.size() > 1) {
+            unusedOperationOutputs.put(field, origins.subList(0, origins.size() - 1));
+          }
+        }
+      }
+
+      unusedOperationOutputs.putAll(redundantOutputs);
+
+      if (!unusedOperationOutputs.isEmpty()) {
+        unusedOutputsFromStage.put(stageName, unusedOperationOutputs);
+      }
+
+      if (!invalidInputs.isEmpty()) {
+        invalidOperationInputs.put(stageName, invalidInputs);
+      }
+    }
+
+    if (unusedOutputsFromStage.isEmpty() && invalidOperationInputs.isEmpty()) {
+      return;
+    }
+
+    String message = invalidOutputErrorMessage(unusedOutputsFromStage);
+    message += invalidInputErrorMessage(invalidOperationInputs);
+    throw new IllegalArgumentException(message);
+  }
+
+  private void processInputOutputs(String operationName, List<String> inputs, List<String> outputs,
```

Comment:
I can't follow the logic because there are so many collections and some of these collections are being mutated so I need to understand how they are changing between method calls. 

Can you refactor this so that it returns what is unused, invalid, and redundant instead of modifying state? If you make it static and package private, it can also be unit tested. Together with a javadoc, it should make it clear what the contract of this method is.